upup/pkg/fi/cloudup/vsphere/vsphere_utils.go

```diff
@@ -0,0 +1,103 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package vsphere
+
+import (
+	"context"
+	"github.com/golang/glog"
+	"github.com/vmware/govmomi/object"
+	"github.com/vmware/govmomi/vim25/mo"
+	"github.com/vmware/govmomi/vim25/types"
+	"path"
+	"sync"
+)
+
+var snapshotLock sync.Mutex
+
+func createSnapshot(ctx context.Context, vm *object.VirtualMachine) (object.Reference, error) {
+	snapshotLock.Lock()
+	defer snapshotLock.Unlock()
+
+	snapshotRef, err := findSnapshot(vm, ctx, snapshotName)
+	if err != nil {
+		return nil, err
+	}
+	glog.V(4).Infof("Template VM is %s and snapshot is %s", vm, snapshotRef)
+	if snapshotRef != nil {
+		return snapshotRef, nil
+	}
+
+	task, err := vm.CreateSnapshot(ctx, snapshotName, snapshotDesc, false, false)
```

Comment:
Defining snapshotName and snapshotDesc in upup/pkg/fi/cloudup/vsphere/vsphere_cloud.go is not serving any purpose. Either pass them as arguments, so that vsphere_utils.go remains independent of these constants, or simply define them here. I would prefer the first option.